### PR TITLE
Fix and improvement of the print methods in SparseMatrix

### PR DESCRIPTION
**Pull request by najlkin**
A fix and an improvement of the print methods in `SparseMatrix`. Specifically, there are two things:

- A fix of the printing when the data are stored on a device, this was ok for the classic `Print`, but other variants (`PrintMatlab`, `PrintMM`, `PrintCSR(2)`) were overlooked.
- An improvement of the methods (`PrintMatlab`, `PrintMM`) to support non-finalized (LIL) matrices like `Print`.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4077](https://github.com/mfem/mfem/pull/4077) | @najlkin | @tzanio | @termi-official + @artv3 | 1/23/24 | 2/1/24 | 2/1/24 | |

<details>
<summary>PR Checklist</summary>
    
- [x] Code builds.
- [x] Code passes `make style`.
- [x] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [x] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [x] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [x] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [x] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [x] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [x] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [x] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by artv3 — APPROVED** [review]

**Comment by najlkin** [comment]
> Can we add some basic regression test for these?

Is there some for `Print()`? Could you point me there so I can just extend that for the other methods.
Or do you mean to manually check if it works as before? I have not changed the original implementation (HostRead does nothing on CPU), so it definitely works the same way on CPU and with finalized matrices.

**Comment by termi-official** [comment]
I think maybe a start is something like this https://github.com/mfem/mfem/blob/0de295d6e1fbfb0442d222e48c8cf95b1fd1d0fe/tests/unit/linalg/test_hypre_vector.cpp#L64-L81

Alternatively you could also print to a stringstream and test the contents.

File: linalg/sparsemat.cpp [hunk 1]
```diff
@@ -3341,6 +3379,10 @@ void SparseMatrix::PrintCSR(std::ostream & os) const
 
    os << height << '\n';  // number of rows
 
+   // HostRead forces synchronization
+   HostReadI();
+   HostReadJ();
+   HostReadData();
```
Comment (jandrej) [inline linalg/sparsemat.cpp:3385]: Isn't this misleading? If you print a CSR matrix you want the state of the data represented in a string. What you do here is a side effect and falsely claims that you did the right thing before.

Comment (najlkin) [reply linalg/sparsemat.cpp:3385]: The actual state used for computation is at the device, so you want to update the host before printing it and not to print an outdated copy. Moreover, I have copied this from `Print`, so it is not a new behavior.

**Review by termi-official — APPROVED** [review]

**Comment by najlkin** [comment]
> I think maybe a start is something like this
> 
> https://github.com/mfem/mfem/blob/0de295d6e1fbfb0442d222e48c8cf95b1fd1d0fe/tests/unit/linalg/test_hypre_vector.cpp#L64-L81
> 
> Alternatively you could also print to a stringstream and test the contents.

That is not possible unfortunately, because `SparseMatrix` does not have methods for loading text files/streams. However, I added a unit test based on that `stringstream` to please you 😉 Have a look.

**Comment by termi-official** [comment]
LGTM.

> > I think maybe a start is something like this
> > https://github.com/mfem/mfem/blob/0de295d6e1fbfb0442d222e48c8cf95b1fd1d0fe/tests/unit/linalg/test_hypre_vector.cpp#L64-L81
> > 
> > Alternatively you could also print to a stringstream and test the contents.
> 
> That is not possible unfortunately, because `SparseMatrix` does not have methods for loading text files/streams. However, I added a unit test based on that `stringstream` to please you 😉 Have a look.

Also totally fine. The code paths should be essentially the same.

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: linalg/sparsemat.cpp (hunk 1 above)
Comment (najlkin) [reply linalg/sparsemat.cpp:3385]: I just added notes to doxygen to let users know 😉 .

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Pull request merged**